### PR TITLE
add helm test functionality 

### PR DESCRIPTION
**Pull request by cqwense**
resolves: https://github.com/Typositoire/concourse-helm3-resource/issues/23

with caveat: I personally don't like the idea of doing both an install/upgrade and then a test via a singular task/step/put in a pipeline - it would make the return code ambiguous - instead test should be run as a on_success , or even as a separate task dependent on install/upgrade -- basically anything except in one asset payload.

--

adds asset/out parameter functionality 

test: true
test_logs: true

inherits existing asset/out functionality:

debug
timeout
namespace
release

### CODE REVIEW

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-1.11.0
+1.11.1
```
Comment (Typositoire) [inline VERSION:1]: That would definitely be a 1.12.0 (New feature being implemented)

File: assets/out [hunk 2]
```diff
@@ -41,11 +42,16 @@ wait=$(jq -r '.params.wait // 0' < $payload)
...
 timeout=$(jq -r '.params.timeout // "5m0s"' < $payload)
 
-if [ -z "$chart" ] && [[ "$delete" == "false" ]]; then
+if [[ "$test" == "false" ]] && [[ "$delete" == "false" ]] && [ -z "$chart" ] ; then
   echo "invalid payload (missing chart)"
   exit 1
 fi
 
+if [[ "$test" == "true" ]] && [[ "$delete" == "true" ]] ; then
+  echo "invalid payload ( test and delete cannot both be true )"
```
Comment (Typositoire) [inline assets/out:51]: Don't tell me what to do :P

File: assets/out [hunk 3]
```diff
@@ -178,7 +184,7 @@ helm_upgrade() {
 
   helm_args=("${upgrade_args[@]}" "${overridden_args[@]}")
   helm_echo_args=("${upgrade_args[@]}" "${scrubbed_overridden_args[@]}")
-  helm_diff_args=("${upgrade_diff_args[@]}" "${overridden_args[@]}" "--suppress-secrets")
+  helm_diff_args=("${upgrade_diff_args[@]}" "${overridden_args[@]}" "--suppress-secrets" "--include-tests" )
```
Comment (Typositoire) [inline assets/out:187]: What happens if there's no tests ?

**Comment by Typositoire** [comment]
I agree that I don't like it being one or the other. Maybe there's something to do with doing tests as part of the "IN" ?

File: assets/out (hunk 3 above)
Comment (cqwense) [reply assets/out:187]: it makes no difference to diff - I was surprised it wasn't included by default , thinking it was just another k8s resource so why specifically have to call it but *shrug* 

but your question did make me wonder what would happen if you ran test on a release with no tests defined, and as it turns out it returns successful, having successfully ran 0 tests.

**Comment by cqwense** [comment]
while I agree there is room for improvement for IN , i don't think IN should run tests as that would be PUT'ing test pods into the k8s cluster.

**Comment by cqwense** [comment]
going to close this while I do some branch merging and resolves , shouldn't be too long.

**Comment by Typositoire** [comment]
> while I agree there is room for improvement for IN , i don't think IN should run tests as that would be PUT'ing test pods into the k8s cluster.

Oh right it creates pods. Ok let's keep it like that then

**Pull request closed without merging**